### PR TITLE
Enable rootless container w/ CI tweaks/bugfixes

### PR DESCRIPTION
**Pull request by theEndBeta**
## Proposed change

Enables running the container as rootless by setting the running user to be the `paperless` user (configurable by `USERMAP_UID`).
Also fixes three other issues:
* building fails on selinux enabled systems unless relabeling is enabled for the bind-mounted host directory
  * `COPY`-ing the files into the container instead works around this problem without requiring relabeling 
* `podman build` fails on non-absolute mount points
  * fixed by  setting the workdir to an argument and reusing that argument in the mount
* builds using github actions failed if the running repository contains any capital letters
  * container registries require all lowercase repo name
  * fixed by ensuring that we coerce the repository name to lowercase

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -68,6 +68,12 @@ jobs:
       - ci-backend
       - ci-frontend
     steps:
+      -
+        name: Set gchr repository name
+        id: set-ghrc-repository
```
Comment (shamoon) [inline .github/workflows/ci.yml:73]: Believe `ghcr` typo x2 here (and same for refs to this)

File: .github/workflows/installer-library.yml [hunk 2]
```diff
@@ -36,6 +36,12 @@ jobs:
     name: Prepare Docker Image Version Data
     runs-on: ubuntu-20.04
     steps:
+      -
+        name: Set gchr repository name
+        id: set-ghrc-repository
```
Comment (shamoon) [inline .github/workflows/installer-library.yml:41]: Same typos

File: docs/setup.rst [hunk 3]
```diff
@@ -184,6 +184,29 @@ Install Paperless from Docker Hub
     port 8000. Modifying the part before the colon will map requests on another
     port to the webserver running on the default port.
 
+    **Rootless**
+
+    If you want to run Paperless as a rootless container, you will need to:
+
+    - set the ``user`` running the container to map to the ``paperless`` user in the
+      container.
+      See ``USERMAP_UID`` and ``USERMAP_GID`` :ref:`here <configuration-polling>`.
```
Comment (shamoon) [inline docs/setup.rst:193]:
```suggestion
      See ``USERMAP_UID`` and ``USERMAP_GID`` :ref:`here <configuration-docker>`.
```

**Comment by shamoon** [comment]
Thanks! Docker / ci isnt my side of the project but I think others have been wanting this! I did notice some small typo-ish things for which I left comments but again, will leave proper review to others. Presumably this supsedes #15

**Comment by shamoon** [comment]
Ps. this looks like it needs a re-base on `dev` as there's a conflict already.

File: docs/setup.rst [hunk 4]
```diff
@@ -184,6 +184,29 @@ Install Paperless from Docker Hub
...
+
+    - override some of the ``supervisord`` defaults by setting the ``command`` to:
```
Comment (bauerj) [inline docs/setup.rst:195]: Please add where this needs to be modified (might not be ovious to everyone).

File: docs/setup.rst [hunk 5]
```diff
@@ -184,6 +184,29 @@ Install Paperless from Docker Hub
...
+
+     .. code::
+
+        command:
+          - "/usr/local/bin/supervisord"
```
Comment (bauerj) [inline docs/setup.rst:200]: This doesn't need to be part of this PR but couldn't we use a wrapper script that sets the appropriate command based on UID?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (theEndBeta) [reply .github/workflows/ci.yml]: Should be fixed

File: .github/workflows/installer-library.yml (hunk 2 above)
Comment (theEndBeta) [reply .github/workflows/installer-library.yml]: Should be fixed

File: docs/setup.rst (hunk 3 above)
Comment (theEndBeta) [reply docs/setup.rst]: Should be fixed

File: docs/setup.rst (hunk 4 above)
Comment (theEndBeta) [reply docs/setup.rst]: Amended. Let me know if there are other changes (or I misunderstood).

**Comment by theEndBeta** [comment]
Added a wrapper script for the `supervisord` docker command.

Also noticed that the bind-mount of the `docker` directory in the `Dockerfile` cannot be read on selinux-enabled systems.
Relabeling by adding `,z` or `,Z` to the end of the mount command will overcome this, but If you don't have to relabel the host filesystem, it's better not to.

File: docs/setup.rst [hunk 6]
```diff
@@ -184,6 +184,25 @@ Install Paperless from Docker Hub
...
+      See ``USERMAP_UID`` and ``USERMAP_GID`` :ref:`here <configuration-docker>`.
+
+    Your entry for Paperless should contain something like:
+
+     .. code::
+
+        webserver:
+          image: ghcr.io/paperless-ngx/paperless-ngx:latest
+          user: <user_id>
+
```
Comment (bauerj) [inline docs/setup.rst:205]: Now that's excellent documentation!

**Review by bauerj — APPROVED** [review]
This looks great to me but I'd still like to hear what @stumpylog thinks about this.

File: Dockerfile [hunk 7]
```diff
@@ -117,22 +117,36 @@ COPY gunicorn.conf.py .
...
+	"docker/paperless_cmd.sh", \
+	"docker/wait-for-redis.py", \
+	"docker/management_script.sh", \
+	"docker/install_management_commands.sh", \
+	"${DOCKER_SRC}" \
+]
 
-RUN --mount=type=bind,readwrite,source=docker,target=./ \
-  set -eux \
+RUN set -eux \
```
Comment (stumpylog) [inline Dockerfile:135]: Presumably this is the change made for SELinux support.  It feels like a lot of change to support what maybe isn't so common?  What's the issue with adding the `z` mount option instead?

File: Dockerfile [hunk 8]
```diff
@@ -211,4 +225,4 @@ ENTRYPOINT ["/sbin/docker-entrypoint.sh"]
 
 EXPOSE 8000
 
-CMD ["/usr/local/bin/supervisord", "-c", "/etc/supervisord.conf"]
+CMD ["/usr/local/bin/paperless_cmd.sh"]
```
Comment (stumpylog) [inline Dockerfile:228]: Nit: I can't say I love this naming for the file, it doesn't feel descriptive to what it does

**Review by stumpylog — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 8 above)
Comment (theEndBeta) [reply Dockerfile:228]: It's the docker `cmd` to run paperless - I wasn't feeling particularly inspired and will heartily accept recommendations.

File: Dockerfile (hunk 7 above)
Comment (theEndBeta) [reply Dockerfile:135]: Why make changes to the host filesystem when you don't have to? As I run mostly selinux systems, I would not appreciate it being done without my permission.

File: Dockerfile [hunk 9]
```diff
@@ -117,22 +117,36 @@ COPY gunicorn.conf.py .
 # setup docker-specific things
 # Use mounts to avoid copying installer files into the image
 # These change sometimes, but rarely
-WORKDIR /usr/src/paperless/src/docker/
+ARG DOCKER_SRC=/usr/src/paperless/src/docker/
```
[the remark this answers is not in the document]
Comment (theEndBeta) [reply Dockerfile:120]: It was required to be `ARG` in a previous changeset (when the following `COPY` was still a bind mount from the host)-
`podman` will not bind mount to a relative directory.

It is also my preference to use absolute paths for the `COPY` destination (which is the other place this `ARG` is used) as it ensures that the command will still do the correct thing if someone decides to reorganize the file.

File: Dockerfile (hunk 7 above)
Comment (stumpylog) [reply Dockerfile:135]: In the end, it's probably just a few kbs of files left in the image, which is sizeable enough that's basically nothing.

**Review by stumpylog — APPROVED** [review]

File: Dockerfile (hunk 7 above)
Comment (theEndBeta) [reply Dockerfile:135]: Adding `rm install_management_commands.sh management_script.sh` at the end of the `RUN` block would accomplish basically the same thing, as those are the only files that would be unused.

Comment (stumpylog) [reply Dockerfile:135]: Once files have been added with COPY, their size always exists at that layer of the image, even if `mv`ed or `rm`ed, thanks to the overlay file system.  But it's just text files here, so not really an issue

**Comment by shamoon** [comment]
Looks like still 1 small conflict to resolve and that’s it!

**Comment by cyruscook** [comment]
Thank You for making this PR, @theEndBeta . I'm a bit confused by the documentation and was wondering if you could clarify some questions I have?
Should the user on the host be "paperless" or something else.
Also, should the "user" variable in the docker compose be formatted like "UID:GID" (e.g. "1000:1000") or just "UID" (e.g. "1000").
And finally, should the UID/GID specified be of the user on the host (e.g. "1000") or the mapping used from /etc/sub(u/g)id (e.g. "100000").
Thanks!

**Comment by theEndBeta** [comment]
> Should the user on the host be "paperless" or something else.

The user on the host can be whatever user you want it to be - the only requirement is that the host users UID has to map the the `paperless` user in the container (whose UID is configurable via `USERMAP_UID`).

> Also, should the "user" variable in the docker compose be formatted like "UID:GID" (e.g. "1000:1000") or just "UID" (e.g. "1000").

Only the UID is necessary, but either should work - the permissions are set for the user.

> And finally, should the UID/GID specified be of the user on the host (e.g. "1000") or the mapping used from /etc/sub(u/g)id (e.g. "100000").

The UID specified in `docker compose`'s `user` or the `--user` flag determines the UID that the process _in the container_ runs as. How that maps back to the host depends on your `userns_mode` setting, which for rootless should default to the `/etc/sub{u,g}id` mapping.

**Comment by Dunuin** [comment]
I still don't understand how to get Paperless-ngx in a rootless docker running.

Got a Debain 11 VM running a rootless Docker with Portainer EE. Docker is run as user "docker" with UID/GID "1000" on the host.
Folders are mounted like this using fstab:
```
# <file system>                                 <mount point>                   <type>          <options>                                                                                       <dump>  <pass>
//192.168.45.4/Paperless_Consumption            /media/Paperless_Consumption    cifs            auto,rw,credentials=/root/.smb_paperless,uid=1000,gid=1000,file_mode=0660,dir_mode=0770         0       0
//192.168.45.4/Paperless_Media                  /media/Paperless_Media          cifs            auto,rw,credentials=/root/.smb_paperless,uid=1000,gid=1000,file_mode=0660,dir_mode=0770         0       0
```
So folders on the host are owned by UID 1000:
```
root@PortainerDMZ:/# ls -la /media/Paperless_Media/
total 0
drwxrwx--- 2 docker docker   0 Aug 21 19:38 .
drwxr-xr-x 8 root   root   131 Aug 21 15:13 ..
drwxrwx--- 2 docker docker   0 Aug 21 19:38 documents
drwxrwx--- 2 docker docker   0 Aug 21 19:20 Export
```
I tried to create a stack with portainer with this compose file...:
```
# docker-compose file for running paperless from the Docker Hub.
# This file contains everything paperless needs to run.
# Paperless supports amd64, arm and arm64 hardware.
#
# All compose files of paperless configure paperless in the following way:
#
# - Paperless is (re)started on system boot, if it was running before shutdown.
# - Docker volumes for storing data are managed by Docker.
# - Folders for importing and exporting files are created in the same directory
#   as this file and mounted to the correct folders inside the container.
# - Paperless listens on port 8010.
#
# In addition to that, this docker-compose file adds the following optional
# configurations:
#
# - Instead of SQLite (default), PostgreSQL is used as the database server.
#
# To install and update paperless with this file, do the following:
#
# - Open portainer Stacks list and click 'Add stack'
# - Paste the contents of this file and assign a name, e.g. 'Paperless'
# - Click 'Deploy the stack' and wait for it to be deployed
# - Open the list of containers, select paperless_webserver_1
# - Click 'Console' and then 'Connect' to open the command line inside the container
# - Run 'python3 manage.py createsuperuser' to create a user
# - Exit the console
#
# For more extensive installation and update instructions, refer to the
# documentation.

version: "3.4"
services:
  broker:
    image: docker.io/library/redis:6.0
    restart: unless-stopped
    volumes:
      - redisdata:/data
    command: redis-server --requirepass MyPassword

  db:
    image: docker.io/library/postgres:13
    restart: unless-stopped
    volumes:
      - pgdata:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: paperless
      POSTGRES_USER: paperless
      POSTGRES_PASSWORD: MyPassword

  webserver:
    image: ghcr.io/paperless-ngx/paperless-ngx:latest
    user: 'docker'
    restart: unless-stopped
    depends_on:
      - db
      - broker
      - gotenberg
      - tika
    ports:
      - 8010:8000
    healthcheck:
      test: ["CMD", "curl", "-fs", "-S", "--max-time", "2", "http://localhost:8000"]
      interval: 30s
      timeout: 10s
      retries: 5
    volumes:
      - data:/usr/src/paperless/data
      - /media/Paperless_Media:/usr/src/paperless/media
      - /media/Paperless_Media/Export:/usr/src/paperless/export
      - /media/Paperless_Consumption:/usr/src/paperless/consume
    environment:
      PAPERLESS_REDIS: redis://:MyPassword@broker:6379
      PAPERLESS_DBHOST: db
      PAPERLESS_TIKA_ENABLED: 1
      PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
      PAPERLESS_TIKA_ENDPOINT: http://tika:9998
# The UID and GID of the user used to run paperless in the container. Set this
# to your UID and GID on the host so that you have write access to the
# consumption directory.
      USERMAP_UID: 1000
      USERMAP_GID: 1000
# Additional languages to install for text recognition, separated by a
# whitespace. Note that this is
# different from PAPERLESS_OCR_LANGUAGE (default=eng), which defines the
# language used for OCR.
# The container installs English, German, Italian, Spanish and French by
# default.
# See https://packages.debian.org/search?keywords=tesseract-ocr-&searchon=names&suite=buster
# for available languages.
      PAPERLESS_OCR_LANGUAGES: deu eng
# Adjust this key if you plan to make paperless available publicly. It should
# be a very long sequence of random characters. You don't need to remember it.
      PAPERLESS_SECRET_KEY: SomePass
# Use this variable to set a timezone for the Paperless Docker containers. If not specified, defaults to UTC.
      PAPERLESS_TIME_ZONE: Europe/Berlin
# The default language to use for OCR. Set this to the language most of your
# documents are written in.
      PAPERLESS_OCR_LANGUAGE: deu
# update checking and notification
      PAPERLESS_ENABLE_UPDATE_CHECK: true
# Postgres password
      PAPERLESS_DBPASS: MyPassword
# how documents should be named in media folder
      PAPERLESS_FILENAME_FORMAT: "{created_year}/{correspondent}/{created_month}-{created_day}_{asn}_{title}"
      PAPERLESS_FILENAME_FORMAT_REMOVE_NONE: false
# autocreate paperless admin user
      PAPERLESS_ADMIN_USER: Admin
      PAPERLESS_ADMIN_MAIL: my@email.com
      PAPERLESS_ADMIN_PASSWORD: MyPassword
# more OCR options
      PAPERLESS_OCR_MODE: skip_noarchive
# setup for 4 vCPUS
      PAPERLESS_TASK_WORKERS: 2
      PAPERLESS_THREADS_PER_WORKER: 2
# scan consumption directory every 300 seconds
      PAPERLESS_CONSUMER_POLLING: 60
# allow to automatically tag documents based on folder they were put in
      PAPERLESS_CONSUMER_RECURSIVE: true
      PAPERLESS_CONSUMER_SUBDIRS_AS_TAGS: true
# set tmp dir to something else as "/tmp" so OCR won't store big documents in tmpfs/RAM
      #PAPERLESS_CONVERT_TMPDIR: /usr/src/paperless/tmp	  
# creation date OCR recognition
      PAPERLESS_DATE_ORDER: DMY
      PAPERLESS_IGNORE_DATES: 01.01.2000

  gotenberg:
    image: docker.io/gotenberg/gotenberg:7.4
    restart: unless-stopped
    command:
      - "gotenberg"
      - "--chromium-disable-routes=true"


  tika:
    image: ghcr.io/paperless-ngx/tika:latest
    restart: unless-stopped

volumes:
  data:
  #media:
  pgdata:
  redisdata:
```
...but it results in a error: `Started Container paperless-ngx-webserver-1 Starting Error response from daemon: unable to find user docker: no matching entries in passwd file`

When removing the line "user: 'docker'" from the compose file it will start the container but the container loops with this message:
```
Paperless-ngx docker container starting...
Installing languages...
Hit:1 http://deb.debian.org/debian bullseye InRelease
Hit:2 http://deb.debian.org/debian-security bullseye-security InRelease
Hit:3 http://deb.debian.org/debian bullseye-updates InRelease
Reading package lists...
Package tesseract-ocr-deu already installed!
Package tesseract-ocr-eng already installed!
Creating directory /tmp/paperless
Adjusting permissions of paperless files. This may take a while.
Waiting for PostgreSQL to start...
Waiting for Redis: redis://:MyPassword@broker:6379
Connected to Redis broker: redis://:MyPassword@broker:6379
Apply database migrations...
/usr/local/lib/python3.9/site-packages/dateparser/date_parser.py:35: PytzUsageWarning: The localize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). For more details on migrating to a PEP 495-compliant implementation, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
  date_obj = stz.localize(date_obj)
SystemCheckError: System check identified some issues:
ERRORS:
?: PAPERLESS_CONSUMPTION_DIR is not writeable
	HINT: Set the permissions of 
drwxrwx--- /usr/src/paperless/consume
 to be writeable by the user running the Paperless services
?: PAPERLESS_MEDIA_ROOT is not writeable
	HINT: Set the permissions of 
drwxrwx--- /usr/src/paperless/media
 to be writeable by the user running the Paperless services
```

So how to get this running?
Other containers just worked fine accessing the SMB shares when telling them to use the UID and GID of the SMB mount.

**Comment by qcasey** [comment]
> I tried to create a stack with portainer with this compose file...:

Without knowing much about this issue, https://github.com/paperless-ngx/paperless-ngx/discussions/887 leads me to think Portainer / CIFS may have something to do with it.

**Comment by Dunuin** [comment]
But does someone know what would be correct values for "USERMAP_UID", "USERMAP_GID" and "user" of the webserver container for a rootless docker in case docker is run as UID/GID 1000 and media/consumption folder are also owned on the host by GID/UID 1000?
Not that I search the problem at portainer/CIFS when my docker-compose is just wrong.

**Comment by theEndBeta** [comment]
@Dunuin Sorry about the delay.

The `user` field in the compose spec specifies the user or uid that the entrypoint will be run by _in the container_. This means that when you specified `user: docker`, it tried to find the nonexistent user `docker` in the container, which is the `passwd file` error you saw.

The `paperless` user is created in the container and thus exists. If `USERMAP_UID`/`USERMAP_GID` are set, the `paperless` user's UID/GID are set to those values. This allows you to know/specify exactly what the UID/GID will be in the container.

You are on the right track with the `subuid`/`subgid` - you may have to experiment to see how docker handles the mappings as I am having a bit of trouble parsing the docker documentation. For podman you would be off by 1 - host UID 1001 would map to container UID 101000, since the calling host UID is mapped to container UID 0.

So, if you set  `USERMAP_UID`/`USERMAP_GID` to `1000`/`1000`, and `user: paperless` (or `user: 1000`, you don't have to use a named user, which would bypass the `passwd file` error) you will have to set the host directory and files to be owned by the mapped uid - `101000` or `100999` depending on how docker actually does the mapping.

**Comment by Dunuin** [comment]
So, so I will try it with UID/GID 101000 and 100999 for the mounted SMB shares and 1000 for "USERMAP_UID", "USERMAP_GID" and "paperless" for the user.

But I still got the problem that I can't set user to "paperless" as then the webserver will fail with a loop of:

```
Paperless-ngx docker container starting...
Installing languages...
Reading package lists...
E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)
Paperless-ngx docker container starting...
Installing languages...
Reading package lists...
E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)
...
```
Without a "user" set it will successfully use apt to search for new OCR languages. Looks like this isn't allowed when setting any "user" parameter so the webserver fails before even trying to check the media/consumption folders.

Edit:
When commenting out the "user: paperless" it won't fail with the "E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)" error. But then it complained about not writable folders when the folders are owned by UID/GID 101000 on the host and "USERMAP_UID", "USERMAP_GID" is set to 1000. Without specifying a user, "USERMAP_UID", "USERMAP_GID" set to 1000 and folders owned by UID/GID 100999 it won't complain any longer about the folders and I finally can access the Paperless-ngx webUI.

Now I will have setup paperless-ngx and see if everything is really working.

So to summarize it as far as I undestand it:
1.) don't use "user: paperless" for the webserver or it will fail because that user isn't allowed to use apt.
2.) not sure which user will be used then to run paperless-ngx inside the container but that user will use UID 1000 and GID 1000 which I defined with "USERMAP_UID: 1000" and "USERMAP_GID: 1000". But as far as I understand the documentation this also wouldn't be needed because UID/GID 1000 is the default paperless-ngx will use.
3.) check your user/group remapping in the files "/etc/subuid" and "/etc/subgid". Here it uses "docker:100000:65536" so UIDs/GIDs 100000 to 165535 on the host will be mapped to 0 to 65535 inside the container. So UID/GID 1000 inside the container should map to UID/GID 101000 on the host. But there is an offset of 1, so actually UID/GID 1000 inside the container will be mapped to 101000 - 1 = UID/GID 100999. So stuff owned by UID/GID 100999 on the host will be owned by UID/GID 1000 inside the guest when bind-mounting folders from host to the docker container.

**Pull request merged**